### PR TITLE
User authentication and mutations

### PR DESCRIPTION
**Pull request by maxmwang**
Implementation of new user authentication design as proposed [here](https://discord.com/channels/853509478142181376/853787947212210206/1077747573357555833). 

# Summary
### Authentication
1. Google OAuth 2.0 with [Passport.js](https://www.passportjs.org)
2. Integration of Express sessions with Apollo resolvers through context
3. Authentication layer as an `@auth` directive

Stays with OAuth 2.0 and does not migrate to OpenID Connect in order to preserve the potential of a future feature: direct integration between Berkeleytime and Google Calendar. Current OAuth scope is limited to profile information, but Google Calendar's API can be added later. 

Migrate from implicit flow to authorization code flow (a summary of both designs can be found [here](https://auth0.com/intro-to-iam/what-is-oauth-2)). Uses [Passport.js](https://www.passportjs.org), an authentication middleware that integrates cleanly with Express, and its [Google OAuth 2.0 strategy](https://www.passportjs.org/packages/passport-google-oauth20/?utm_source=github&utm_medium=referral&utm_campaign=passport-google-oauth20&utm_content=about). 

### Mutations
1. Added/removed properties from `User` schema
    [–] `password`
    [–] `is_superuser`
    [–] `date_joined` (replaced with createdAt, automatically created with timestamps)
    [+] `google_id`
    [+] `refresh_token`
3. Added [SchemaType options](https://mongoosejs.com/docs/schematypes.html#schematype-options)
5. Added mutations (all require user authentication):
    * `UpdateUserInfo`
    * `DeleteUser`

## Authentication Implementation
Primarily done under Passport.js's hood, the design flow is outlined as follows:
1. Frontend user clicks login button, which sends a GET request to backend's `/login` endpoint
2. Backend redirects user to a Google OAuth login page
6. Frontend user logs into Google account, sending a GET request to backend's `/login/callback` with an authorization code stored in a query string
7. Backend verifies code through Google API, redirecting frontend user on login success/failure and creating a user session

If it's the user's first time logging in, a new User object will be created and saved into MongoDB. To logout, POST `/logout`.

## Sessions
The user session is stored in `req.user` and is the same as the User object stored in Mongo. The object contains sensitive data (accessToken, refreshToken) and should **NOT** be sent back to the frontend. This session is integrated into Apollo through [context](https://www.apollographql.com/docs/apollo-server/data/context/). 

# Use in Backend
Authentication is done through an `@auth` directive, which serves as a sort of middleware between the apolloServer and the resolver. Right now, only valid locations are `OBEJCT` and `FIELD_DEFINITION`. An example:

```ts
// src/modules/user/resolver.ts

const typedef = gql`
  directive @auth on OBJECT | FIELD_DEFINITION // <-------- is in the common module's typedef

  type User { ... }

  type Query @auth { // <-------- OBJECT
    User: User
  }

  type Mutation {
    UserChangeFirstName(first_name: String!): User @auth // <-------- FIELD_DEFINITION
  }
`;
```

The `@auth` directive will ensure the user is authenticated before the resolver is called. 

To access context in a resolver, add it as a parameter to the resolver function:

```ts
// src/modules/user/resolver.ts
const resolvers: UserModule.Resolvers = {
    Query: {
        User:(_, __, contextValue: any) {
            // use contextValue.user, ie:
            console.log("Id: ", contextValue.user._id);

            return getUserById(contextValue.user._id);
        }),
    },
};
```

# To Do in Backend
1. Update `.env` with `URL`, `GOOGLE_CLIENT_ID`, `GOOGLE_CLIENT_SECRET`, `SESSION_SECRET`
2. Add authorized origin and redirect url to Google Cloud console 
3. Configure session cookie maxAge (determines after how long will a user have to relogin)

# To Do in Frontend
1. Rewrite LogIn component to send GET request to `/auth/google`
5. Enable axios withCredentials option with `axios.defaults.withCredentials = true`

### CODE REVIEW

**Review by kevinzwang — CHANGES_REQUESTED** [review]
Great work! LGTM so far. Thanks for the detailed comments too! Can you implement a resolver in the code that uses user auth so it could be tested out before we merge?

One more comment: the `secure` wrapper function paradigm seems novel to me. Could you talk about how you decided to go with it?

**Comment by maxmwang** [comment]
> Great work! LGTM so far. Thanks for the detailed comments too! Can you implement a resolver in the code that uses user auth so it could be tested out before we merge?

Yeah of course! It's used a lot in [`user-mutations`](https://github.com/asuc-octo/berkeleytime-gql/tree/user-mutations) right now and seems to be working fine. I'll just add `secure` to the query in this branch. 

> One more comment: the `secure` wrapper function paradigm seems novel to me. Could you talk about how you decided to go with it?

The biggest problem was that we want to add this auth layer to specific endpoints—specific resolvers. So, an umbrella layer (which would likely live in our `express.ts` loader) that encompasses all endpoints would not work (although this would be a very clean solution). Instead, we need a way to target individual resolvers, and so the auth layer must live in each `resolver.ts`. However, as stated before, this layer has little to do with the actual resolvers, and so intuitively should not live in the `resolver.ts` files. 

The `secure` wrapper function serves as a way to target individual resolvers without creating too much bloat code within the `resolver.ts` files.

**Comment by kevinzwang** [comment]
Have you considered using custom directives? That allows us to define authenticated queries and datatypes in our GraphQL schema instead of in our resolvers, so we can specify auth for not just queries but also specific fields or types within a query. As a bonus result it would also visible to the client via a schema query. 

- [Apollo Server docs on auth with custom directives](https://www.apollographql.com/docs/apollo-server/security/authentication#with-custom-directives)
- [graphql-tools schema directives docs](https://the-guild.dev/graphql/tools/docs/schema-directives)

**Comment by maxmwang** [comment]
Updated original PR comment with `@auth` directive documentation.

File: src/bootstrap/graphql/directives/auth.ts
Comment (kevinzwang) [inline src/bootstrap/graphql/directives/auth.ts:1]: looks fine to me but either I or someone else should review [the docs](https://the-guild.dev/graphql/tools/docs/schema-directives) and make sure this logic makes sense.

File: src/bootstrap/loaders/passport.ts
Comment (kevinzwang) [inline src/bootstrap/loaders/passport.ts:1]: I'm going to trust that this is secure and uses best practices but also may be worth having someone who knows better check.

One question I do have though: if a user deletes their account and then recreates it, will their original session token still work? it might be worth having user-level session invalidation, which we can also serve as a mutation

File: src/config.ts
Comment (kevinzwang) [inline src/config.ts:1]: seems kind of redundant to define the type to use it a single time. ik it was here already but do we need to do this?

File: src/modules/user/formatter.ts
Comment (kevinzwang) [inline src/modules/user/formatter.ts:1]: lol can we just cast `user as UserModule.User` instead of writing this?

File: src/modules/user/resolver.ts
Comment (kevinzwang) [inline src/modules/user/resolver.ts]: is email the best identifier for a user? just making sure cuz it may be plausible that you can change your email address but have the same account.

File: src/modules/user/typedefs/user.ts
Comment (kevinzwang) [inline src/modules/user/typedefs/user.ts]: adding @auth here actually makes all queries require auth. put them on each individual query instead. same for mutations below

File: src/modules/user/typedefs/user.ts
Comment (kevinzwang) [inline src/modules/user/typedefs/user.ts]: field names should be lowercase. same for mutations below

File: src/modules/user/typedefs/user.ts
Comment (kevinzwang) [inline src/modules/user/typedefs/user.ts]: Is there a reason why these two mutations are not part of `updateUserInfo`? the parameters are all optional anyway. you can create [input types](https://graphql.org/graphql-js/mutations-and-input-types/) to group the parameters for cleanliness

File: src/modules/user/model.ts
Comment (kevinzwang) [inline src/modules/user/model.ts:1]: please move this to `src/db` to match master

**Review by kevinzwang — CHANGES_REQUESTED** [review]

File: src/modules/user/typedefs/user.ts
Comment (maxmwang) [reply src/modules/user/typedefs/user.ts]: Wouldn't basically all mutations require `@auth` for Users? If so, wouldn't it be repetitive to put it on each individual mutation instead of just all mutations?

Comment (kevinzwang) [reply src/modules/user/typedefs/user.ts]: I don't anticipate any non-authenticated mutations, so if you think so as well, we could keep that on the Mutation type. But if you are going to do that, u should put it in the `common` module instead. In fact you should put the directive definition in `common` too.

Do set it on a per-field basis for queries tho.

File: src/modules/user/controller.ts
Comment (kevinzwang) [inline src/modules/user/controller.ts]: `key` should be `value`

**Comment by zacharyzollman** [comment]
Is `user._id`  just ObjectID? How will you know that value prior to retrieving user data via the query? It's true that email addresses can change, so should we have the option to query by Google ID?

**Comment by maxmwang** [comment]
> Is `user._id` just ObjectID? How will you know that value prior to retrieving user data via the query? It's true that email addresses can change, so should we have the option to query by Google ID?

The `_id` is inside the user session object stored in `contextValue.user`. If the user is logged, then they won't need to know their `_id` to query themselves.

File: src/bootstrap/loaders/passport.ts
Comment (maxmwang) [reply src/bootstrap/loaders/passport.ts:1]: > One question I do have though: if a user deletes their account and then recreates it, will their original session token still work? it might be worth having user-level session invalidation, which we can also serve as a mutation

Fixed so that on user deletion, user session cookie is invalidated. Not too sure if this will also invalidate the entry in Redis.

File: src/modules/user/resolver.ts
Comment (maxmwang) [reply src/modules/user/resolver.ts]: Switched to MongoDB's ObjectID

**Pull request merged**
